File: v1/java/trinary/src/main/java/Trinary.java
public class Trinary {
    private static final int INVALID_TRINARY_DECIMAL_VALUE = 0;

    public static int toDecimal(String input) {
        return calculateDecimalValue(input);
    }

    private static int calculateDecimalValue(String input)
    {
        if (isValidTrinaryString(input)) {
            return input.chars()
                    .reduce(0, (acc, digit) -> acc * 3 + digit - '0');
        }

        return INVALID_TRINARY_DECIMAL_VALUE;
    }

    private static boolean isValidTrinaryString(String input) {
        return input.matches("^[012]*$");
    }
}

File: v1/ecmascript/trinary/trinary.js
export default class Trinary {
    constructor(trinary) {
        this.digits = /^[012]*$/.test(trinary) ? [...trinary] : [];
    }

    toDecimal() {
        return this.digits.reduce((acc, c) => acc * 3 + this.charToDecimal(c), 0)
    }

    charToDecimal(c) {
        return c - '0';
    }
}